KAFKA-10259: KIP-554 Broker-side SCRAM Config API

rondagostino opened the pull request:
I still need to add a test file `core/src/test/scala/unit/kafka/admin/UserScramCredentialsCommandTest.scala`, but aside from that I think it's close to done.

rondagostino committed:
KAFKA-10259: KIP-554 Broker-side SCRAM Config API

rondagostino committed:
Test cases for no credentials and authz failures

rondagostino committed:
ConfigCommand illgal invocation tests

rondagostino committed:
add negative test for --entity-default

rondagostino committed:
Describe works, HMAC_SHA -> SCRAM_SHA

rondagostino committed:
more negative tests

rondagostino committed:
some AlterUserScramCredentials tests

rondagostino committed:
create/describe/delete/describe test

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1214,6 +1215,60 @@ default AlterClientQuotasResult alterClientQuotas(Collection<ClientQuotaAlterati
+    /**
+     * <p>Describe SASL/SCRAM credentials.
+     *
+     * @param users the users for which credentials are to be described; all users' credentials are described if null or empty
```
It would be good to document here that users that are specified but not  present will be silently omitted from the result

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/ScramMechanism.java:
```diff
@@ -0,0 +1,70 @@
+     * @see <a href="https://tools.ietf.org/html/rfc5802#section-4>
+     *     Salted Challenge Response Authentication Mechanism (SCRAM) SASL and GSS-API Mechanisms, Section 4</a>
+     */
+    public static ScramMechanism fromMechanismName(String mechanismName) {
```
Hmm... for consistency with `from(byte)`, shouldn't we return UNKNOWN if we don't know the name?

cmccabe commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -980,4 +984,137 @@ class AdminManager(val config: KafkaConfig,
+          val propertyValue = userConfig.getProperty(mechanism.toMechanismName)
+          if (propertyValue != null) {
+            val iterations = ScramCredentialUtils.credentialFromString(propertyValue).iterations
+            userScramCredentials.credentialInfos.add(new CredentialInfo().setMechanism(mechanism.ordinal.toByte).setIterations(iterations))
```
I'd really prefer some way to just get the `value` byte out of the ScramMechanism object, rather than relying on the ordering of enum entries.  If we are going to rely on the ordering, that would have to be prominently documented in the class (but let's not....)

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4071,6 +4081,168 @@ void handleFailure(Throwable throwable) {
+            UserScramCredentialDeletion deletion = (UserScramCredentialDeletion) alteration;
+            ScramMechanism mechanism = deletion.getMechanism();
+            if (mechanism == null || mechanism == ScramMechanism.UNKNOWN) {
+                userIllegalAlterationExceptions.put(deletion.getUser(), new IllegalArgumentException("Unknown SCRAM mechanism"));
```
We should have a defined ApiException and Errors entry for errors like this.  We could maybe use InvalidConfigurationException / INVALID_CONFIG

rondagostino replied:
How about `InvalidRequestException`?  It's already used in this class, and it might be more appropriate than `InvalidConfigurationException`.

rondagostino replied:
I think `UnsupportedSaslMechanismException` is actually appropriate here -- it already exists and corresponds to `Errors.UNSUPPORTED_SASL_MECHANISM`.

cmccabe replied:
ok

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4071,6 +4081,168 @@ void handleFailure(Throwable throwable) {
+                        // we might overwrite aa exception from a previous upsertion, but we don't really care
+                        // since we just needs to mark this user as having at least one illegal alteration
+                        // and make an exception instance available for completing the corresponding future exceptionally
+                        userIllegalAlterationExceptions.put(user, e);
```
We should have a defined ApiException subclass for this.  NoSuchAlgorithmException could probably be shoehorned into INVALID_CONFIG.  InvalidKeyException probably needs its own defined error code (which we should add to the KIP).

This may seem bureaucratic, but think about it this way: maybe we'll want to do this on the broker side eventually.  Then we need a way to serialize, which plain old java exceptions don't have.

rondagostino replied:
I wonder if `InvalidRequestException` is indeed the right one.  On the broker side, in `kafka.server.AdminManager`, the code looks like this:

```
    invalidUsers.foreach(user => retval.results.add(new AlterUserScramCredentialsResult().setUser(user)
      .setErrorCode(Errors.INVALID_REQUEST.code).setErrorMessage("Unknown SCRAM mechanism or too few iterations")))
```

cmccabe replied:
The pattern in Kafka has always been to have very specific error codes and exceptions.  Unfortunately INVALID_REQUEST is documented like this:
```
"This most likely occurs because of a request being malformed by the " +
"client library or the message was sent to an incompatible broker. See the broker logs " +
"for more details.",
```
So we probably need a new error code here

rondagostino replied:
`NoSuchAlgorithmException` is thrown if we don't support the (otherwise valid) SCRAM mechanism.  For example, this might occur if we add support for a new SCRAM mechanism but the client is running an older JRE that doesn't support the algorithm associated with that newer mechanism.  `UnsupportedSaslMechanismException` would be appropriate here, I think.

`InvalidKeyException` is thrown if the password to be salted is unacceptable -- i.e. if it is empty.  I will add a new UnacceptablePasswordException and Errors.UNACCEPTABLE_PASSWORD as a starting point for further review/discussion.

rondagostino replied:
Actually, we also have to deal with too few/too many iterations as well as empty username.  These, along with empty password, are a class unto themselves -- unacceptable credentials.  So I'm abandoning the specific "`UnacceptablePasswordException`/`Errors.UNACCEPTABLE_PASSWORD`" in favor of "`UnacceptableCredentialException`/`Errors.UNACCEPTABLE_CREDENTIAL`"

cmccabe commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -365,45 +368,90 @@ object ConfigCommand extends Config {
-        alterEntityTypes.zip(alterEntityNames).foreach { case (k, v) => alterEntityMap.put(k, v) }
-        val entity = new ClientQuotaEntity(alterEntityMap)
+        if (hasQuotaConfigsToAdd || hasQuotaConfigsToDelete) {
+          // handle altering client/user quota configs
```
it would be good to split this off into its own function (after verifying that we weren't trying to do both SCRAM + quota)

cmccabe commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -365,45 +368,90 @@ object ConfigCommand extends Config {
+          adminClient.alterClientQuotas(Collections.singleton(new ClientQuotaAlteration(entity, alterOps)), alterOptions)
+            .all().get(60, TimeUnit.SECONDS)
+        } else {
+          // handle altering user SCRAM credential configs
```
it would be good to split this off into its own function (after verifying that we weren't trying to do both SCRAM + quota) or multiple entity names

cmccabe commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -980,4 +984,137 @@ class AdminManager(val config: KafkaConfig,
+      }
+    }
+
+    if (users.isEmpty)
```
It would be good to avoid confusing empty with null.  Why not use Option[Seq[String]]

rondagostino committed:
Respond to review comments

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4071,6 +4081,168 @@ void handleFailure(Throwable throwable) {
+        final KafkaFutureImpl<Map<String, UserScramCredentialsDescription>> future = new KafkaFutureImpl<>();
+        final long now = time.milliseconds();
+        Call call = new Call("describeUserScramCredentials", calcDeadlineMs(now, options.timeoutMs()),
+                new ControllerNodeProvider()) {
```
Does this need to be controller for Describe? It looks similar to DescribeAcls where we use LeastLoadedNode.

rondagostino replied:
The KIP says `It will be will be sent to the controller, and will return NOT_CONTROLLER if the receiving broker is not the controller.` It says this for both Describe and Alter.  I agree it doesn't seem necessary for Describe.  Would it require an email to the list for notification if we decide to change it?  Or would a KIP update be sufficient?  Do you think we should change it?  @cmccabe any thoughts here as well?

rajinisivaram replied:
Let's see what @cmccabe thinks. We can do whatever we intend to do for similar describe APIs in the KIP-500 world. If we are changing, then we should update the KIP and send a note to the KIP discussion thread.

cmccabe replied:
We're planning on getting rid of ControllerNodeProvider as part of KIP-590, so I think these should all just use `leastLoadedNode`.  Right now they set / fetch stuff in ZK directly anyway.

Yes, it will require an email to the mailing list, but I don't think it's a big change

rondagostino replied:
Ok, switching Describe to **not** require that it be done on the controller.

> planning on getting rid of ControllerNodeProvider as part of KIP-590 

Leaving Alter alone for now under the assumption that we will fix it as part of the KIP-590 PR.  Let me know if you wish me to change this now instead.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4071,6 +4081,168 @@ void handleFailure(Throwable throwable) {
+        return new AlterUserScramCredentialsRequestData.ScramCredentialDeletion().setName(d.getUser()).setMechanism(d.getMechanism().getType());
+    }
+
+    private static byte[] getSaltedPasword(ScramMechanism publicScramMechanism, byte[] password, byte[] salt, int interations) throws NoSuchAlgorithmException, InvalidKeyException {
```
typo: iterations

rajinisivaram commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -508,7 +563,15 @@ object ConfigCommand extends Config {
+    // we describe user SCRAM credentials only when we are not describing client information
+    // and we are not given either --entity-default or --user-defaults
+    if (!entityTypes.contains(ConfigType.Client) && !entityNames.contains("")) {
+      getUserScramCredentialConfigs(adminClient, entityNames).foreach { case (user, description) =>
```
What do we do if the user is not a SCRAM user? Won't this throw an exception? Can we make sure that user without quota or SCRAM credential doesn't print any errors or exceptions?

rondagostino replied:
I'm adding `core/src/test/scala/unit/kafka/admin/UserScramCredentialsCommandTest.scala` -- let me know if this test covers this case.

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -980,4 +984,137 @@ class AdminManager(val config: KafkaConfig,
+        retval.setError(Errors.INVALID_REQUEST.code())
+        retval.setErrorMessage(s"Cannot describe SCRAM credentials for the same user twice in a single request: ${duplicatedUsers.mkString("[", ", ", "]")}")
+      } else
+        users.get.foreach { user => addToResults(user, adminZkClient.fetchEntityConfig(ConfigType.User, Sanitizer.sanitize(user))) }
```
should we catch exception?

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -980,4 +984,137 @@ class AdminManager(val config: KafkaConfig,
+      }).map(_.name) ++ deletions.filter(deletions => deletions.name.isEmpty || scramMechanism(deletions.mechanism) == ScramMechanism.UNKNOWN).map(_.name))
+      .toSet
+    invalidUsers.foreach(user => retval.results.add(new AlterUserScramCredentialsResult().setUser(user)
+      .setErrorCode(Errors.INVALID_REQUEST.code).setErrorMessage("Unknown SCRAM mechanism or too few iterations")))
```
Can't we return an exception message that says exactly why, instead of `x or y`

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -980,4 +984,137 @@ class AdminManager(val config: KafkaConfig,
+    val initiallyValidUserMechanismPairs = (upsertions.filter(upsertion => !invalidUsers.contains(upsertion.name)).map(upsertion => (upsertion.name, upsertion.mechanism)) ++
+      deletions.filter(deletion => !invalidUsers.contains(deletion.name)).map(deletion => (deletion.name, deletion.mechanism)))
+
+    // https://stackoverflow.com/questions/24729544/how-to-find-duplicates-in-a-list
```
Do we include this kind of comment in code?

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -980,4 +984,137 @@ class AdminManager(val config: KafkaConfig,
+          val publicScramMechanism = scramMechanism(upsertion.mechanism)
+          publicScramMechanism == ScramMechanism.UNKNOWN ||
+            (upsertion.iterations != -1 &&
+              InternalScramMechanism.forMechanismName(publicScramMechanism.getMechanismName).minIterations > upsertion.iterations)
```
Shouldn't we also limit the max value for iterations?

rajinisivaram replied:
Shouldn't we also limit the max value for iterations?

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2973,6 +2975,40 @@ class KafkaApis(val requestChannel: RequestChannel,
+    if (!controller.isActive) {
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        alterUserScramCredentialsRequest.getErrorResponse(requestThrottleMs, Errors.NOT_CONTROLLER.exception))
+    } else if (authorize(request.context, ALTER, CLUSTER, CLUSTER_NAME)) {
```
I think we should also not allow users who authenticated using delegation tokens to create or update users. We don't allow these users to create new tokens, it would be odd if they could create a new user or the password of the user of the token.

rajinisivaram reviewed (commented):
@rondagostino Thanks for the PR, looks good. I reviewed the implementation (not tests) and left some comments.

rondagostino committed:
Respond to review, add UserScramCredentialsCommandTest.scala

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1214,6 +1215,64 @@ default AlterClientQuotasResult alterClientQuotas(Collection<ClientQuotaAlterati
+     *
+     * @param users the users for which credentials are to be described; all users' credentials are described if null
+     *              or empty.  A user explicitly specified here that does not have a SCRAM credential will not appear
+     *              in the results.
```
Should we throw an exception for users which don't exist to be consistent with other APIs?

rondagostino replied:
Hmm, good question.  The KIP doesn't state what do do here.  @cmccabe thoughts?

cmccabe replied:
I originally was thinking we could just omit them.  But thinking about it more, I think we should have an error code to be consistent with our other APIs.  We'll need a new error code like USER_NOT_FOUND.  And we'll also need to add a per-user error code-- something we don't currently have.  But it shouldn't be too hard to do.

cmccabe replied:
(Another advantage to having a per-user error code is that in the future, we may have more reasons why describing a user might fail, and these might be per-user)

cmccabe replied:
@rondagostino : It still says "A user explicitly specified here that does not have a SCRAM credential will not appear in the results".  I thought we agreed that such a user would get an error code?

rondagostino replied:
I reworked this Javadoc to list all possible exceptions for both alter and describe.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/ScramMechanism.java:
```diff
@@ -0,0 +1,81 @@
+     * @see <a href="https://tools.ietf.org/html/rfc5802#section-4>
+     *     Salted Challenge Response Authentication Mechanism (SCRAM) SASL and GSS-API Mechanisms, Section 4</a>
+     */
+    public String getMechanismName() {
```
We don't use `get` prefix elsewhere, just `mechanismName()`?

rondagostino replied:
I made the changes here and also removed all `get` prefixes on getters within `ScramCredentialInfo` and `UserScramCredential{Alteration,Deletion,Upsertion,Description}`

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/ScramMechanism.java:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return the type indicator for this SASL SCRAM mechanism
+     */
+    public byte getType() {
```
As before, `type()`?

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/UserScramCredentialsDescription.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return the user name
+     */
+    public String getName() {
```
Remove `get` prefix

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/UserScramCredentialsDescription.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return the unmodifiable list of SASL/SCRAM credential representations for the user
+     */
+    public List<ScramCredentialInfo> getInfos() {
```
`credentialInfos()`?

rajinisivaram commented on core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala:
```diff
@@ -0,0 +1,387 @@
+  }
+
+  @Test
+  def testAlterEmptyUser(): Unit = {
```
Do we test for empty password?

rondagostino replied:
Added a test for it in `UserScramCredentialsCommandTest`.  We can't test for it here because we get a salted password here, and I don't think it is possible for that to be an empty string and allow a successful SASL/SCRAM authentication.

rajinisivaram commented on core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala:
```diff
@@ -0,0 +1,387 @@
+      val results = response.data.results
+      assertEquals(2, results.size)
+      assertTrue("Expected error when altering the same credential twice in a single request",
+        results.get(0).errorCode == Errors.INVALID_REQUEST.code && results.get(1).errorCode == Errors.INVALID_REQUEST.code)
```
assertEquals so we know what the error was if test fails? (mutliple places)

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/ScramCredentialInfo.java:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     *
+     * @return the number of iterations used when creating the credential
```
or -1?

rondagostino replied:
TBD.  No change needed if we are getting rid of `-1` as a special value.

rondagostino committed:
Leverage admin client in tests to create SCRAM credentials

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4071,6 +4081,168 @@ void handleFailure(Throwable throwable) {
+
+    private static byte[] getSaltedPasword(ScramMechanism publicScramMechanism, byte[] password, byte[] salt, int iterations) throws NoSuchAlgorithmException, InvalidKeyException {
+        return new ScramFormatter(org.apache.kafka.common.security.scram.internals.ScramMechanism.forMechanismName(publicScramMechanism.getMechanismName()))
+                .hi(password, salt, iterations);
```
Iterations can be -1 here? Won't we end up sending a password without applying the salt properly?

rondagostino replied:
Yeah, I think there is a fundamental problem with allowing the value `-1` for iterations.  The KIP says "`Note that if the number of iterations is set to -1, the server-side default will be used.`". But we are on the client at this point in the code, and there is no concept for "server-side default" for SCRAM iterations in Kafka.  And unfortunately since we are salting the password here, we need to know the number of iterations.  So I think we need to do either of the following:

1. Add an ability to define server-side default number of iterations per SASL/SCRAM mechanism in Kafka and allow clients to learn them.
2. Get rid of `-1` as a special value.

It's pretty clear to me that (2) is the way to go, but @rajinisivaram and @cmccabe please chime in.

rajinisivaram replied:
Agree that we should just get rid of -1.

rajinisivaram reviewed (commented):
@rondagostino  Left a few more comments. I was about to ask if we can add an end-to-end test, but looks like you have just added those. The reason I was interested in the end-to-end test was because I wasn't sure if -1 for iteration count where broker chooses iteration count actually works. Have we tested that?

rondagostino committed:
Respond to review comments

rondagostino reviewed (commented):
Thanks for the reviews, @rajinisivaram .  Take a look and let me know what you think -- especially about the "server-side default iterations" value of -1.

rondagostino committed:
Use admin client in system tests

rondagostino committed:
Eliminate -1 as a valid iterations value

rondagostino commented:
@cmccabe and @rajinisivaram Thanks for the reviews.  I think this is where we stand:

1. I removed `-1` as a valid `iterations` value in the code.  I will need to announce this in the DISCUSS email thread and adjust the KIP.  I will do that once we have the full set of KIP changes based on all these points.
2. We decided that we will not allow users who authenticated using delegation tokens to alter user SRAM credentials.  I will add a mention of this to the KIP once we have the full set of KIP changes based on all these points.
2. The KIP says that Describe `will be sent to the controller, and will return NOT_CONTROLLER if the receiving broker is not the controller.`. This constraint is perhaps not necessary for Describe requests.  If we are all in agreement then I can change the code, and I will adjust the KIP once we have the full set of KIP changes based on all these points.
3. We have to decide what to do if a client asks to Describe specific users that don't exist.  The KIP does not explicitly say what to do in this situation, but the Describe Response only has a top-level error; unlike the Alterations response, the Describe response doesn't provide an error per user.  So the entire Describe request currently has to succeed or fail as a whole.  Therefore the code currently just silently ignores users that don't exist and only describes users that do exist (i.e. if you request to describe `user1`, `user2`, and `user3`, but `user2` doesn't exist (or it does exist but doesn't have any SCRAM credentials), the response indicates success and describes just `user1` and `user3`.  We need to decide if the current behavior is acceptable, and if it is acceptable, I would add a line to the KIP making this behavior explicit (since I just implied it based on the single error design).  If this behavior is not acceptable, then I think we would need to adjust the format of the response, the code, and the KIP.  Thoughts?
4. I updated existing both the `integration tests` and `system tests` to use the admin client when creating User SCRAM credentials after the brokers start instead of going directly to ZooKeeper; ZooKeeper is still used for bootstrapping broker credentials prior to starting Kafka, of course.  These changes haven't been reviewed yet, I don't think, and we should probably kick off test runs to exercise everything (tests pass locally for me).

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,50 @@
+ */
+@InterfaceStability.Evolving
+public class DescribeUserScramCredentialsResult {
+    private final KafkaFuture<Map<String, UserScramCredentialsDescription>> future;
```
If we're going to have a per-user error, then we need `Map<String, KafkaFuture<...>>`.  This will also be useful if we need to add more per-user errors in the future.

rondagostino replied:
Now that I'm working on this, I discovered that there is no other API that can describe or list everything that works this way.  Everything that can describe or list everything returns a single future.  Every describe or list API that returns a map of keys to futures requires a non-empty list of keys to describe or list.  For example:

1. `listTopics()` lists all topics and returns a single `Future`; the `describeTopics()` API returns a map of names to futures but requires a non-empty list of topics to describe.
2. `describeConfigs()` returns a map of resources to futures but requires a non-empty list of resources to describe.
3. `describeLogDirs()` returns a map of broker IDs to futures but requires a non-empty list of brokers to describe.
4. `describeReplicaLogDirs()` returns a map of replicas to futures but requires a non-empty list of replicas to describe.
5. `describeConsumerGroups()` returns a map of consumer groups to futures but requires a non-empty list of consumer groups to describe.
6. `listPartitionReassignments()` allows listing all or a subset of reassignments and returns a single future.
7. `listOffsets()` returns a map of topic-partitions to futures but requires a non-empty list of topic-partitions to describe.
8. `describeClientQuotas()` allows listing all or a subset of quotas and returns a single future.

I think if we made this change here we would be off the beaten path.  That's not necessarily bad, but what tipped me off to this was the fact that when we list everything we have to create a future for every user that gets returned, and we don't know that list of users when we make the request, so there's really no way to implement it.

We could create two separate APIs: one for describing some explicit, non-empty list of users, which would return a map of users to futures, and another one that describes everything, which returns a single future.  `listTopics()` vs `describeTopics()` works this way, for example, though the information returned in the two is very different: when listing you just get the names, and when describing you get a lot more.  I don't see us distinguishing between listing vs. describing in terms of data -- we are going to send back the same two things (mechanism and iterations) regardless.  So we would probably be talking about creating a `describeUserScramCredentials()` API and a `describeAllUserScramCredentials()` API with the first taking a list and returning a map of futures and the second not taking a list and returning a single future.

But I'm thinking we should just keep it the way it is -- take a possibly-empty list and return a single future regardles of whether the list was empty or not.

Thoughts?

rondagostino replied:
I don't see the need for multiple futures; everything is returning in a single response from Kafka, and I can't think of why it would be necessary to have some users' credential descriptions available to the API caller but others not.  I'll take a look at keeping a single future for the response but moving the error status from the top level down into each per-user result.

rondagostino replied:
I kept the top-level error information but added per-user error information in addition.

cmccabe replied:
Thinking about this more, I would prefer having both a list RPC to show everything, and a describe RPC to show only some things.  There are a few reasons why.  One is that even though we currently only make one RPC, in the future we might make more than one.  In that case we would want multiple futures.

Another is that the general pattern in Kafka is that list RPCs show everything, but describe RPCs show only some things.  It's true that there are some places where we violate this pattern, but I still think it's worth trying to follow where we can.  Maybe this should be documented better so that when we add new RPCs, people aren't confused about whether to use "list" or "describe."

I also feel like in AdminClient, errors should be handled with futures pretty much all the time, unless there is a really strong reason not to.  This allows people to use an async style of programming.  In contrast, mixing in some errors that aren't futures, but need to be checked explicitly is very likely to confuse people.

rondagostino replied:
> a list RPC to show everything, and a describe RPC to show only some things.

Do you mean a list RPC that takes no arguments and returns every credential defined for every user and a describe RPC that takes 1 or more users and returns every credential defined for those specified users, and they both return the same information for each credential?

Or do you mean a list RPC and a describe RPC that return different sets of information (as is done with list vs. describe topics)?  I think you mean the former (two RPCs, each returning the same thing), but I want to be certain I understand.

> There are a few reasons why. One is that even though we currently only make one RPC, in the future we might make more than one. In that case we would want multiple futures.

I don't understand what this is referring to.  By "we currently only make one RPC" to what are you referring?

> I also feel like in AdminClient, errors should be handled with futures pretty much all the time

Agreed.  Will convert to using futures always, whenever we arrive at the final decision on what the RPCs need to look like.

I'm wondering if we convert to returning futures everywhere, can we stick with the one describe RPC?  For example, could the admin client return a `Future<Map<String, Future<UserScramCredentialDescription>>>`?  Would that work, and if so, would that be a reasonable way to proceed?

rondagostino replied:
If we do decide to go with 2 separate APIs, then I might be concerned about using `list()` vs `describe()` if they return the same set of information (i.e. mechanism and iterations).  Perhaps using two separate names gives us room to expand `describe()` to return more information later on, though.  But if not, and they will always return the same information, then maybe `describeAll()` and `describe()` (or `listAll()` and `list()`) might be better?

rondagostino replied:
I can see reasonable arguments for all the possibilities, so if you feel strongly about one of them then I would be fine with it.  For example, `list()` and `describe()` even though they return the same thing (now -- `describe()` could return more later potentially).  Just let me know.

cmccabe replied:
I thought about this more and I think I see a good way out of this difficulty.  We should just have an accessor method like `userInfo(String name)` that returns a `KafkaFuture<user info>`.  We can dynamically create this future if needed.  Then we can have a single RPC and a single API which is just `describe`

rondagostino replied:
Ok, I think I follow now.  You are saying that we could potentially implement describe by invoking 1+N requests to Kafka: one to get the list of credentials (either the list of all of them if we are asking for them all, or the explicitly requested ones we wanted), and then another N requests to get the data for each one.  This on the surface seems like an anti-pattern, but it is not unreasonable for the case where the data is expensive to get in the first place — maybe we are forced to make 1 or more round-trips for each anyway.  So as a general, reusable pattern, yes, I believe it works.

So when we invoke describe, whether it is describe-all or just an explicit few, we return a single future, and that future returns a list of instances (UserName in this case): either 1 instance for each user that has at least 1 credential for the describe-all case, or one instance per distinct user explicitly requested otherwise.  Then each UserName instance has the accessor you mentioned, which in this case returns  Future<UserScramCredentialDescription>.

Do I have that right?

rondagostino replied:
Ok, we have two levels of Futures now:

1. `DescribeUserScramCredentialsResult` has a `KafkaFuture<List<UserScramCredentialsDescriptionResult>>`
2. `UserScramCredentialsDescriptionResult` has a user name and a `KafkaFuture<UserScramCredentialsDescription>`

I think this is where we want to end up.  Let me know if you agree.  I also added some logic in `AlterUserScramCredentialsRequestTest` to confirm the behavior of the per-user futures and associated errors.

cmccabe commented on core/src/test/scala/unit/kafka/admin/UserScramCredentialsCommandTest.scala:
```diff
@@ -0,0 +1,127 @@
+    val utf8 = StandardCharsets.UTF_8.name
+    val printStream = new PrintStream(byteArrayOutputStream, true, utf8)
+    var exitStatus: Option[Int] = None
+    Exit.setExitProcedure { (status, _) =>
```
How does this interact with JUnit running tests concurrently?  cc @ijuma @hachikuji

rondagostino replied:
Hmm, good point, I think there may be a problem here in general because there is only a single exit procedure that can be set globally, and multiple tests that set/reset it in parallel will collide.  There are 16 Scala test classes in `core` out of 260 that do this -- so 6% of test classes.  So I think this will introduce some flakiness to these 16 tests.  Does this sound correct to you, and should we open a separate ticket for this as opposed to trying to fix it here?

rondagostino replied:
Actually, I think this may not be an issue since parallel tests in Gradle run in separate processes rather than separate threads.  From https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html: `"Test are always run in (one or more) separate JVMs."`

cmccabe commented on core/src/test/scala/unit/kafka/server/DescribeUserScramCredentialsRequestTest.scala:
```diff
@@ -0,0 +1,128 @@
+}
+
+object DescribeCredentialsTest {
+  val UnauthorizedPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "Unauthorized")
```
It's kind of a code smell to be mutating static data during a JUnit test.  There are ways to run JUnit tests in parallel (although I think our current setup only runs at the granularity of test classes, not test methods?)  At the very least we'd have to document that this is not thread-safe and that it will prevent us from parellelizing this test class in the future.

Rather than doing this, how about having two separate test authorizers, and choosing the one you want for each individual JUnit test method?

In order to fit this into our bass-akwards inheritance based test design ( :( ) you could check the name of the test in `brokerPropertyOverrides` using one of these methods: https://stackoverflow.com/questions/473401/get-name-of-currently-executing-test-in-junit-4

rondagostino replied:
@cmccabe Thanks.  I copied this approach from `kafka.server.RequestQuotaTest`.  I fixed it here in `DesrcibeUserScramCredentialsRequestTest` and `AlterUserScramCredentialsRequestTest` but was unable to fix it in `RequestQuotaTest` because that test is checking every possible API, and of course certain APIs need to succeed for Kafka to start successfully.  So that test needs to mutate the behavior after `setUp()`.  These tests don't since these APIs aren't necessary for Kafka to start.  So I was able to fix it here but was unable to fix it where I got the approach from.

cmccabe commented on core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala:
```diff
@@ -0,0 +1,396 @@
+  val UnauthorizedPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "Unauthorized")
+  // Principal used for all client connections. This is modified by tests which
+  // check unauthorized code path
+  var principal = KafkaPrincipal.ANONYMOUS
```
Same comment applies here as in the describe test: don't mutate static data.

cmccabe commented on clients/src/main/java/org/apache/kafka/common/errors/ResourceNotFoundException.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class ResourceNotFoundException extends ApiException {
+    private final String resource;
+
```
We've always set `serialVersionUID` in exception classes, to appease the Java gods.
```
    private static final long serialVersionUID = 1L;
```
it's not really necessary since we don't plan on using Java's built-in serialization framework, but I think some tools complain if they don't see it.  And anyway, it's the pattern... for now.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/UserScramCredentialUpsertion.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private static byte[] generateRandomSalt() {
+        return new BigInteger(130, new SecureRandom()).toString(Character.MAX_RADIX).getBytes(StandardCharsets.UTF_8);
```
I think we can avoid a trip through BigInteger with something like this:
```
      SecureRandom random = new SecureRandom();
      byte bytes[] = new byte[length];
      random.nextBytes(bytes);
```

rondagostino replied:
@cmccabe  I think the approach you suggest leaves out how to identify `length` which itself needs to be randomized.  I got the current implementation from `org.apache.kafka.common.security.scram.internals.ScramFormatter`.  I would have invoked `ScramFormatter.secureRandomBytes()` directly, but it is not `static` and I did not want to either instantiate an instance or change methods to static (though the class is internal, so I could have done that). I instead replicated the logic here.  The array length ends up being random with this approach, as do the bytes in the array.  Let me know what you think.  Currently I've left this as-is.

rondagostino replied:
The latest commit changes the methods on `org.apache.kafka.common.security.scram.internals.ScramFormatter` to be static, and I now reuse that logic.

rondagostino committed:
Respond to review except for per-user errors on Describe

rondagostino committed:
Add per-user errors to DescribeUserScramCredentialResponse

rondagostino commented:
@cmccabe Thanks again for the reviews.  I think all comments are addressed.  Here's where we stand.  I will make the indicated KIP changes and announce them in the DISCUSS email thread once we have the full set of KIP changes agreed to/reviewed in this PR.

1. I removed -1 as a valid iterations value in the code. 
2. Users who authenticate using delegation tokens are not allowed to alter user SRAM credentials.
3. It is no longer necessary to send Describe requests to the controller
4. I added per-user error information in the Describe response, and kept the top-level error information.  The top-level error is non-zero if there is an authorization issue or there is some kind of infrastructure issue.  Otherwise, if some users can't de described but others can, the top-level error indicates success and the per-user errors indicate success or failure accordingly.  Asking to describe a user that doesn't exist or asking to describe the same user twice results in the corresponding user-level error being set.
5. We've added 3 new errors/exception classes: `RESOURCE_NOT_FOUND/ResourceNotFoundException`, `DUPLICATE_RESOURCE/DuplicateResourceException`, and `UNACCEPTABLE_CREDENTIAL/UnacceptableCredentialException` and we leverage those in the expected places.
6. We leverage the existing `UNSUPPORTED_SASL_MECHANISM/UnsupportedSaslMechanismException` where it would be expected.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/UserScramCredentialsDescription.java:
```diff
@@ -0,0 +1,89 @@
+     * @return the always non-null/unmodifiable list of SASL/SCRAM credential representations for the user
+     * (empty if {@link #exception} defines an exception)
+     */
+    public List<ScramCredentialInfo> credentialInfos() {
```
It would be good to throw the exception here if there is one, so that it wasn't possible to ignore the problem

cmccabe commented on clients/src/test/java/org/apache/kafka/common/protocol/ProtoUtilsTest.java:
```diff
@@ -26,11 +26,18 @@
+                    assertTrue(key + " should require delayed allocation", key.requiresDelayedAllocation);
+                    break;
+                default:
+                    assertFalse(key + " should not require delayed allocation", key.requiresDelayedAllocation);
```
I know it's not strictly necessary, but it would be nice to have a "break" after the default clause too

cmccabe commented on core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala:
```diff
@@ -486,7 +486,9 @@ class ConfigCommandTest extends ZooKeeperTestHarness with Logging {
-  def shouldNotAlterNonQuotaClientConfigUsingBootstrapServer(): Unit = {
+  def shouldNotAlterNonQuotaNonScramUserOrClientConfigUsingBootstrapServer(): Unit = {
+    // when using --bootstrap-server, it should be illegal to alter anything that is not a quota and not a SCRAM credential
+    // for both user and client entities
```
sorry, this might be a silly question, but how are these constraints different when using --zookeeper?  should we test that as well

rondagostino replied:
@cmccabe Good question, actually.  There is already a check to make sure a non-existent config cannot be **deleted** via `--zookeeper`: `shouldNotUpdateConfigIfNonExistingConfigIsDeletedUsingZookeper()`.  This test passes, of course.

However, there is no check to make sure an unrecognized config can be **added**, and in fact if I add that test it fails; the code will gladly go ahead and add anything we wish (and it will gladly go ahead and delete it if we wish as well -- the above test is only checking that something that doesn't exist can't be deleted).

The next question, of course, is whether we should "fix" this or not.  What do you think?  To fix it we would need the full set of allowed configs at the User, Client, Topic, and Broker levels and then insert code to check accordingly.  Since the ZooKeeper update path is going away due to KIP-500, I'm wondering if we can just leave it alone?

cmccabe replied:
Yeah, I think we should leave it alone for now, for the reasons you state.  But thanks for the context.

rondagostino committed:
throw exception, add break statement as per review

rondagostino committed:
Return two levels of Future in describe

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,50 @@
+     *
+     * @return the future representing the result of the call
+     */
+    public KafkaFuture<List<UserScramCredentialsDescriptionResult>> future() {
```
Based on our offline discussion, I thought we were going to have three methods here:
```
KafkaFuture<Map<String, UserScramCredentialsDescription>> all()
```
Returns all the users that were listed, or throws an exception if there was an error describing any of them.

```
KafkaFuture<List<String>> users();
```
Returns the names of all the users that were described, whether the describing was successful or not.

```
KafkaFuture<UserScramCredentialDescription> description(String userName);
```
Returns a future describing userName, or RESOURCE_NOT_FOUND if the user was not listed or not found.

rondagostino replied:
Sorry, my mistake.  This now reflects what we discussed.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/UserScramCredentialsDescription.java:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-554%3A+Add+Broker-side+SCRAM+Config+API">KIP-554: Add Broker-side SCRAM Config API</a>
+ */
+public class UserScramCredentialsDescription {
```
As I commented above, I think we should not have the exception here.

rondagostino committed:
Eliminate UserScramCredentialsDescriptionResult

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,139 @@
+            if (requestedUserFuture == null) {
+                throw new ResourceNotFoundException("No such user: " + userName);
+            }
+            KafkaFuture<Void> succeedsOnlyIfRequestedUserFutureSucceeds = KafkaFuture.allOf(requestedUserFuture);
```
We don't need this extra future, right?  We can just use requestedUserFuture directly if we want to use that

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,139 @@
+
+    private static <T> T valueFromFutureGuaranteedToSucceedAtThisPoint(KafkaFuture<T> future) {
+        try {
+            return future.get();
```
You should not be calling `get()` here.

In general it seems like what you really want is a single future behind the scenes that returns everything that you fetched from your single RPC.  Then you just need a few translation functions that pull out the part you need for each future that you're returning to the user.

rondagostino replied:
>  a single future behind the scenes that returns everything that you fetched from your single RPC 

That would introduce an assumption that none of the users' results are available until all of the users' results are available.  For example, currently this class accepts two arguments in its constructor: a future to a list of user names, and a map of usernames-to-futures for the results of each user.  If instead for the second argument we accept a single future to a map of usernames-to-results, then that future must succeed all the time (because we need to distinguish between successes and failures for different users).  Each user's result would have to contain the error code, and no user's results will be available to the client until all of the users' results (the underlying map) is available.  It is true that this is how it is implemented today -- no user's results will be available until all users' results are available, and each of the individual results has an error code -- but Is it acceptable to impose such an all-or-nothing result availability constraint?  The idea of flexibility in the API going forward is an ambiguous one, so we should explicitly decide here whether to keep it the way it is now (allowing different users to complete at different times even though it isn't possible right now) vs. changing the constructor and constraining the results for any particular user to be available only after the results for all users are available.  What do you think?

rondagostino replied:
> You should not be calling get() here

Looking at the implementation of `KafkaFutureImpl` I agree we should not be calling `get()` because I think it means when the users future completes, the thread that completes it will wait until the results future for that user also completes.  If this is indeed the case then we definitely do not want to be calling `get()` in our code.  One way to make a hard dependency on the users future completing without causing the admin client thread to block as just described is to provide a 2-level results hierarchy as was previously implemented.  This forces the user to complete 2 futures and to do it in their own thread.

I wonder if we have the constructor accept 2 futures rather than 1+N futures we would still create a dependency between the 2 that would block the admin client thread.  Perhaps the second future would have to already have a dependency on the users future built into it.

One thing we have to be careful about is the constraints we put on the arguments to the constructor -- those constraints, whatever they end up being, have to be clearly documented.  If we don't have to impose any -- then that's great.  But if we have to impose/document some, then we should compare that with the 2-level results hierarchy I had provided earlier -- that imposes no constraints whatsoever.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     *
+     * @return a future indicating the distinct users that were requested (either explicitly or implicitly via
```
I think it would be better to make this "a future indicating the users that were listed" (rather than "requested").  It's maybe a bit of a subtle distinction but think about things like requesting the null user, or the empty string user.  It's awkward to put that here.  I think if we explicitly request a user but it doesn't exist, it should be omitted from here as well.  That gives us more flexibility in the future with the API.

rondagostino replied:
We've gone back and forth on this.  The KIP does not explicitly state what to do in the case of a describe request for a user that does not have credentials, and we originally coded it to silently drop them, but then we changed it to be consistent with other APIs and raise an error (https://github.com/apache/kafka/pull/9032#discussion_r468871453).  I agree that it isn't totally clear what to do.  Rather than making the change back, I'll leave both this Javadoc and the underlying implementation as-is right now unill we discuss further and decide for sure what we want.

cmccabe commented:
@rajinisivaram : I'd like to understand your suggestion to forbid authenticating via delegation token here.  It doesn't seem consistent with how we handle delegation tokens in general, so I might be missing something.  It seems like a lot of administrative systems may use delegation tokens and this would make this API not useful for them.

rondagostino committed:
Eliminate unnecessary intermediate future as per review

rondagostino committed:
Provide 1 Future to DescribeUserScramCredentialsResult

rondagostino committed:
Javadoc fix

rondagostino committed:
users() returns described users

rondagostino committed:
Added DescribeUserScramCredentialsResultTest

cmccabe commented on tests/kafkatest/services/security/security_config.py:
```diff
@@ -276,18 +287,19 @@ def setup_node(self, node):
-    def setup_credentials(self, node, path, zk_connect, broker):
-        if broker:
-            self.maybe_create_scram_credentials(node, zk_connect, path, self.interbroker_sasl_mechanism,
+    def setup_credentials(self, node, path, connect, creating_broker_user):
```
I think we should just have a separate function for this, since there's no code shared between when creating_broker_user=True and when it is False.

cmccabe commented on tests/kafkatest/services/security/security_config.py:
```diff
@@ -350,6 +362,14 @@ def kafka_opts(self):
         else:
             return ""
 
+    def export_kafka_opts_for_admin_client_as_broker(self):
```
As we discussed offline, I think eventually the admin client should log in as its own admin user, not as the broker.  But we can save that for a follow-on.

rondagostino replied:
Yes, created https://issues.apache.org/jira/browse/KAFKA-10443 to track this.

cmccabe commented on tests/kafkatest/services/kafka/kafka.py:
```diff
@@ -575,10 +577,21 @@ def set_unclean_leader_election(self, topic, value=True, node=None):
+        # Use this for everything related to kafka-configs except User SCRAM Credentials
         if node.version.kafka_configs_command_uses_bootstrap_server():
-            return "--bootstrap-server %s " % self.bootstrap_servers(self.security_protocol)
+            return "--bootstrap-server %s --command-config <(echo '%s')" % (self.bootstrap_servers(self.security_protocol),
```
It's sort of weird that we're creating files on the fly with `'<(echo '%s')` rather than just using a file in `/mnt/security`.  This will result in some pretty long command lines, right?

Considering this is an existing pattern in `kafka.py` (but nowhere else?), let's file a follow-on JIRA to look into this and fix it.  Unless there's some really good reason why `kafka.py` is doing this, but I can't think of any.

rondagostino replied:
I created https://issues.apache.org/jira/browse/KAFKA-10451 to track this.

cmccabe commented on tests/kafkatest/version.py:
```diff
@@ -63,8 +63,13 @@ def reassign_partitions_command_supports_bootstrap_server(self):
 
+    def kafka_configs_command_uses_bootstrap_server_scram(self):
+        # User SCRAM Credentials (KIP-554)
+        return self > LATEST_2_6
```
I don't think this is quite right.  It should be `self >= V_2_7_0`, right?  And we should add 2.7 to the `version.py` file.  

master branch is already identifying itself as 2.7 anyway.

```
[cmccabe@zeratul kafka2]$ git grep 2.7.0 | tail -n 10
docs/upgrade.html:<h5><a id="upgrade_270_notable" href="#upgrade_270_notable">Notable changes in 2.7.0</a></h5>
gradle.properties:version=2.7.0-SNAPSHOT
kafka-merge-pr.py:DEFAULT_FIX_VERSION = os.environ.get("DEFAULT_FIX_VERSION", "2.7.0")
streams/quickstart/java/pom.xml:        <version>2.7.0-SNAPSHOT</version>
streams/quickstart/java/src/main/resources/archetype-resources/pom.xml:        <kafka.version>2.7.0-SNAPSHOT</kafka.version>
streams/quickstart/pom.xml:    <version>2.7.0-SNAPSHOT</version>
streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:     * @deprecated Since 2.7.0; use {@link #addGlobalStore(StoreBuilder, String, Consumed, ProcessorSupplier)} instead.
streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/StreamsBuilder.scala:    "2.7.0"
tests/kafkatest/__init__.py:__version__ = '2.7.0.dev0'
tests/kafkatest/version.py:DEV_VERSION = KafkaVersion("2.7.0-SNAPSHOT")
```

Similarly `supports_tls_to_zookeeper` should be checking for 2.5 and later, not "after 2.4" (they should be the same, but what if someone messed up incrementing LATEST_2_4, etc....)  It's always better to be explicit about what version something appeared in than to say it happened "after (but not including) some version X"

cmccabe commented on core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala:
```diff
@@ -169,6 +169,18 @@ object JaasTestUtils {
+  // Returns a SASL/SCRAM configuration using credentials for the given user and password
+  def scramClientLoginModule(mechanism: String, scramUser: String, scramPassword: String): String = {
+    mechanism match {
+      case "SCRAM-SHA-256" | "SCRAM-SHA-512" =>
```
Can we not have these be naked constants?  Maybe we could iterate over an enum or something?  If we add another mechanism, it would be nice not to have to manually edit this.

rondagostino replied:
I fixed this in the other places in this file as well

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,145 @@
+     * descriptions complete successfully.
+     */
+    public KafkaFuture<Map<String, UserScramCredentialsDescription>> all() {
+        return KafkaFuture.allOf(dataFuture).thenApply(v -> {
```
There's no reason to call `allOf` on a single future.  After all, `allOf`'s function is convert multiple futures to a single one.  But if you already have a single future, this is not needed.

You could use `dataFuture.thenApply`, but that will not trigger if `dataFuture` is completed exceptionally.

Instead, what you want here is something like the following:
`dataFuture.whenComplete( (data, exception) -> if (exception != null) { ... error ... }  else { ... handle data ... }`

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java:
```diff
@@ -0,0 +1,145 @@
+
+    private static <T> T valueFromFutureGuaranteedToSucceedAtThisPoint(KafkaFuture<T> future) {
+        try {
+            return future.get();
```
The call to `get()` here isn't needed.  See the comment about about using `KafkaFuture#whenComplete`.  Chaining future results is a big part of the power of `CompletableFuture` (which `KafkaFuture` is a clone of... long story...)

cmccabe commented on core/src/test/scala/unit/kafka/server/ClientQuotasRequestTest.scala:
```diff
@@ -17,6 +17,8 @@
 
 package kafka.server
 
+import java.util
```
This should go with the other java imports, which I guess are combined in this file

cmccabe commented on core/src/test/scala/unit/kafka/server/ClientQuotasRequestTest.scala:
```diff
@@ -166,14 +168,11 @@ class ClientQuotasRequestTest extends BaseRequestTest {
-    val configs = adminZkClient.fetchEntityConfig(entityType, userName)
-    configs.setProperty(mechanism.mechanismName, ScramCredentialUtils.credentialToString(credential))
-    adminZkClient.changeConfigs(entityType, userName, configs)
+    val results = createAdminClient().alterUserScramCredentials(util.Arrays.asList(
```
You need to close the admin client after creating it.  I don't think Scala has try-with-resources, so you can just use a `try`... `finally` block.

cmccabe commented on core/src/test/scala/unit/kafka/admin/UserScramCredentialsCommandTest.scala:
```diff
@@ -0,0 +1,135 @@
+      }
+      ConfigCommandResult(byteArrayOutputStream.toString(utf8))
+    } catch {
+      case e: Exception => {
```
is it useful to log the exception here at debug level?

rondagostino replied:
Logging it at debug level doesn't hurt, so I added it.

cmccabe commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -1047,8 +1047,8 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
-    createScramCredentials(zkConnect, JaasTestUtils.KafkaScramAdmin, JaasTestUtils.KafkaScramAdminPassword)
+    createScramCredentials(adminClients.head, JaasTestUtils.KafkaScramUser, JaasTestUtils.KafkaScramPassword)
+    createScramCredentials(adminClients.head, JaasTestUtils.KafkaScramAdmin, JaasTestUtils.KafkaScramAdminPassword)
     initializeKerberos()
```
Is it necessary to wait for the change to be applied on all brokers after completing the admin client call?

rondagostino replied:
Good point.  It wasn't waiting before, and it probably didn't/doesn't matter since we were spending time initializing Kerberos, but I added the check anyway just to be safe.

cmccabe commented on core/src/test/scala/integration/kafka/api/SaslScramSslEndToEndAuthorizationTest.scala:
```diff
@@ -42,7 +42,18 @@ class SaslScramSslEndToEndAuthorizationTest extends SaslEndToEndAuthorizationTes
+    createScramCredentialWithScramAdminClient(JaasTestUtils.KafkaScramUser2, JaasTestUtils.KafkaScramPassword2)
+  }
+
+  private def createScramCredentialWithScramAdminClient(user: String, password: String) = {
```
So the reason for not using the `SaslSertup#createScramCredentials` method here is because we want the admin client itself to be authenticated with SCRAM?

rondagostino replied:
It was a goal to eliminate all SCRAM credential creation via ZooKeeper where possible.  The only places that do so after this PR are when credentials have to be created before the brokers are started (i.e. when the inter-broker security protocol is SASL/SCRAM).  This code used to create the credential directly via ZooKeeper, but since it occurs after the brokers start it can use the admin client.

cmccabe commented on core/src/test/scala/integration/kafka/api/SaslClientsWithInvalidCredentialsTest.scala:
```diff
@@ -248,4 +250,25 @@ class SaslClientsWithInvalidCredentialsTest extends IntegrationTestHarness with
     createProducer()
   }
+
+  private def createScramAdminClient(user: String, password: String): Admin = {
```
It feels like we are accumulating a lot of these "create an admin client, but with SCRAM" functions.  Since all these tests ultimately subclass SaslSetup, can't we have a common function there?

rondagostino replied:
Ok, I added the following code to SaslSetup, and we implement that first method in the 3 test classes that use this functionality.

```
  def createPrivilegedAdminClient(): Admin = {
    // create an admin client instance that is authorized to create credentials
    throw new UnsupportedOperationException("Must implement this if a test needs to use it")
  }

  def createScramCredentialsViaPrivilegedAdminClient(userName: String, password: String): Unit = {
    val privilegedAdminClient = createPrivilegedAdminClient() // must explicitly implement this method
    try {
      // create the SCRAM credential for the given user
      createScramCredentials(privilegedAdminClient, userName, password)
    } finally {
      privilegedAdminClient.close()
    }
  }
```

cmccabe commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -545,6 +558,16 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
     }
   }
 
+  protected def createScramAdminClient(scramMechanism: String, user: String, password: String): Admin = {
```
can we create a common function in SaslSetup for this?  Seems to be repeated in a lot of tests.

rondagostino replied:
See above

rondagostino replied:
Ok, this now invokes a new method on SaslSetup().

rondagostino committed:
Respond to review

rondagostino committed:
comment fixes

rondagostino committed:
Create SaslSetup.createAdminClient() method

rondagostino committed:
Allow creation of scram credentials with delegation token

cmccabe commented:
ok to test

cmccabe commented:
ok to test

rondagostino committed:
Fix core spotbugs error under JDK 8/Scala 2.12

rondagostino commented:
System test run was successful except for 2 known flaky tests.  https://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2020-09-02--001.1599074927--rondagostino--KAFKA-10259--b7d1fee5a/report.txt

```
passed:           512
failed:           2

test_id:    kafkatest.tests.connect.connect_distributed_test.ConnectDistributedTest.test_bounce.connect_protocol=compatible.clean=True
status:     FAIL

test_id:    kafkatest.tests.core.replication_test.ReplicationTest.test_replication_with_broker_failure.security_protocol=SASL_SSL.failure_mode=clean_shutdown.broker_type=leader
status:     FAIL
```

rondagostino commented:
Test failures are flakes/unrelated to this PR, no overlaps across JDKs.

cmccabe commented:
LGTM
